main.go

```diff
@@ -561,7 +564,7 @@ func (c *controller) sync(ctx context.Context) {
 		time.Sleep(c.options.scaleTimeout) // Give some time for all replicas before they receive hundreds req/s
 	}
 
-	c.populate(hashrings, statefulsets)
+	c.populate(ctx, hashrings, statefulsets)
```

Comment:
Going through some of this code, my understanding is that there are a few problems that this is addressing in one go,

- In case of first controller `sync`, we weren't checking pod status at all, but just populating hashring.
- In case of scale down, we weren't checking pod status for the remaining or terminating pods, but just removing scaled down ones.
- In case of scale up, we were only checking the status of newly added pods, but not of the older ones due to `sync` replica check, and then adding them.

Doing the pod status in `populate` seems like the best approach here as it accounts for all cases (readiness/terminating) and removes whatever node is down regardless of scenario. 

I think we were always changing hashring via `populate` anyways but were just checking pod status on scale-up only? This would've manifested as disruptions for hashmod and some errors for ketama.

However, I wonder if we can shave down some of the complexity here. 
WDYT about having a `--minimize-hashring-change` flag, which makes sure that we only call `populate` for a hashring, if there is a scale up i.e `c.replicas[hashring] < *sts.Spec.Replicas`, and we always poll and check pod status in `populate` but not in `sync`? 
And the default behavior would be the same as `--allow-only-ready-replicas` and `--allow-dynamic-scaling` being always true. 

That way we can recommend `--minimize-hashring-change` flag for hashmod setups, and for _any_ hashring change, we always check all pod status before modifying. 🙂 

Feel free to correct me if I'm wrong somewhere!

Reply:
Why instead of recommending `--minimize-hashring-change` for hashmod we don't just use it by default on this scenario?

Reply:
Like, auto-detect from StatefulSet/ConfigMap if hashring is configured with hashmod and use the options? Yes that is also a way!

Reply:
Precisely. I think many things become simpler if we can detect the hashring algorithm that is supplied through the statefulset/configmap and automatically turn on the best feature to run with it. 

For the sake of simplicity and step-by-step change, we can say we don't support (yet) the feature where different tenants can use different hashring algorithms.